{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CS 124: Machine Learning in Genetics\n",
    "# Project: Haplotype Phaser\n",
    "# Contributors: Aditya Pimplaskar, Aditya Joglekar\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools as it\n",
    "# Packages needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some helpers\n",
    "def fillna(col):\n",
    "    if col.value_counts().index[0] == '1':\n",
    "        col.fillna(col.value_counts().index[1], inplace=True) # ensure we don't fill heterozygous\n",
    "    else:\n",
    "        col.fillna(col.value_counts().index[0], inplace=True)\n",
    "    return col\n",
    "\n",
    "def imputeData(df):\n",
    "    df = df.replace('*', np.NaN)\n",
    "    #df = df.astype('int64')\n",
    "    #imputer = SimpleImputer(missing_values=np.nan, strategy= 'most_frequent')\n",
    "    #return pd.DataFrame(imputer.fit_transform(df))\n",
    "    return df.apply(lambda col:fillna(col))\n",
    "\n",
    "def readAndImpute(file):\n",
    "    data = pd.read_csv(file, sep = \" \", header = None)\n",
    "    data = imputeData(data)\n",
    "    data = data.astype('int64')\n",
    "    return data.values.tolist()\n",
    "\n",
    "def basic_phaser(genotype):\n",
    "    h1, h2 = [], []\n",
    "    for g in genotype:\n",
    "        if g == 0:\n",
    "            h1.append(0)\n",
    "            h2.append(0)\n",
    "        if g == 2:\n",
    "            h1.append(1)\n",
    "            h2.append(1)\n",
    "        if g == 1:\n",
    "            break\n",
    "            h1.append(0)\n",
    "            h2.append(1)\n",
    "    return h1, h2\n",
    "\n",
    "def add_haplos(h1, h2):\n",
    "    #print(h1, h2)\n",
    "    a = [h1[i] + h2[i] for i in range(len(h1))]\n",
    "    return a # assume h1 and h2 same len\n",
    "\n",
    "def difference(g,h):\n",
    "    d = [g[i]-h[i] for i in range(len(g))]\n",
    "    return d\n",
    "\n",
    "def valid_haplo(h):\n",
    "    for s in h:\n",
    "        if s != 0 and s != 1:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "def h2_from_g_and_h1(g, known):\n",
    "    for h1 in known:\n",
    "        h2 = difference(g, h1)\n",
    "        if valid_haplo(h2):\n",
    "            return h1, h2\n",
    "    return [],[]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "ex1 = readAndImpute(\"assignment/example_data_2_masked.txt\")\n",
    "# ex1 is currently a list of lists\n",
    "# 1 list per SNP, each list has 50 entries (that individual's status at that SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clarks(genotypes):\n",
    "    numSNPS = len(genotypes)\n",
    "    numIndivs = len(genotypes[0])\n",
    "    \n",
    "    \n",
    "    # need to get pool of known haplotypes\n",
    "    def fill_known_haps(df):\n",
    "        nSNPs = len(df)\n",
    "        nIndividuals = len(df[0])\n",
    "        haplotypes = np.zeros((nSNPs, 2*nIndividuals), dtype = np.int)\n",
    "        haplotypes.fill(-1)\n",
    "        known = []\n",
    "        for i in range(nIndividuals):\n",
    "            genotype = [row[i] for row in df]\n",
    "            h1, h2 = basic_phaser(genotype)\n",
    "            if len(h1) == len(h2) == len(genotype):\n",
    "                if h1 not in known:\n",
    "                    known.append(h1)\n",
    "                if h2 not in known:\n",
    "                    known.append(h2)\n",
    "                for SNP in range(nSNPs):\n",
    "                    haplotypes[SNP][2*i] = h1[SNP]\n",
    "                    haplotypes[SNP][2*i + 1] = h2[SNP]\n",
    "        return haplotypes, known\n",
    "    \n",
    "    haplotypes, known = fill_known_haps(genotypes)\n",
    "    \n",
    "    def hash_haplo_combos(known):\n",
    "        combos = {}\n",
    "        for pair in list(it.combinations(known, 2)):\n",
    "            combos[str(add_haplos(pair[0], pair[1]))] = [pair[0],pair[1]]\n",
    "        return combos \n",
    "    \n",
    "    known_combos = hash_haplo_combos(known)\n",
    "    for i in range(7):\n",
    "        for i in range(numIndivs):\n",
    "            genotype = [row[i] for row in genotypes]\n",
    "            genotype_str = str(genotype)\n",
    "            if genotype_str in known_combos:\n",
    "                h1, h2 = known_combos[genotype_str]\n",
    "                for index in range(numSNPS):\n",
    "                    haplotypes[index][2*i] = h1[index]\n",
    "                    haplotypes[index][2*i+1] = h2[index]\n",
    "            else:\n",
    "                h1, h2 = h2_from_g_and_h1(genotype, known)\n",
    "                if len(h1) > 0 and len(h2) > 0:\n",
    "                    for index in range(numSNPS):\n",
    "                        haplotypes[index][2*i] = h1[index]\n",
    "                        haplotypes[index][2*i+1] = h2[index]\n",
    "                    known.append(h2)\n",
    "                    known_combos = hash_haplo_combos(known)\n",
    "\n",
    "    haplotypes = guessUnphased(genotypes, haplotypes)\n",
    "    return haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "row 0 is 0\n",
      "row 0 is 2\n",
      "row 0 is 2\n",
      "row 0 is 2\n",
      "row 1 is 1\n",
      "row 1 is 2\n",
      "row 1 is 2\n",
      "row 1 is 2\n",
      "row 2 is 2\n",
      "row 2 is 1\n",
      "row 2 is 2\n",
      "row 2 is 2\n",
      "row 3 is 2\n",
      "row 3 is 1\n",
      "row 3 is 1\n",
      "row 3 is 2\n",
      "row 4 is 1\n",
      "row 4 is 1\n",
      "row 4 is 2\n",
      "row 4 is 2\n",
      "row 5 is 1\n",
      "row 5 is 2\n",
      "row 5 is 1\n",
      "row 5 is 2\n",
      "row 6 is 2\n",
      "row 6 is 1\n",
      "row 6 is 2\n",
      "row 6 is 2\n",
      "row 7 is 2\n",
      "row 7 is 2\n",
      "row 7 is 2\n",
      "row 7 is 2\n",
      "row 8 is 1\n",
      "row 8 is 2\n",
      "row 8 is 2\n",
      "row 8 is 2\n",
      "row 9 is 2\n",
      "row 9 is 2\n",
      "row 9 is 1\n",
      "row 9 is 2\n",
      "row 10 is 2\n",
      "row 10 is 2\n",
      "row 10 is 2\n",
      "row 10 is 2\n",
      "row 11 is 1\n",
      "row 11 is 2\n",
      "row 11 is 1\n",
      "row 11 is 2\n",
      "row 12 is 1\n",
      "row 12 is 2\n",
      "row 12 is 1\n",
      "row 12 is 2\n",
      "row 13 is 2\n",
      "row 13 is 1\n",
      "row 13 is 2\n",
      "row 13 is 1\n",
      "row 14 is 1\n",
      "row 14 is 2\n",
      "row 14 is 1\n",
      "row 14 is 1\n",
      "row 15 is 1\n",
      "row 15 is 2\n",
      "row 15 is 2\n",
      "row 15 is 2\n",
      "row 16 is 0\n",
      "row 16 is 2\n",
      "row 16 is 1\n",
      "row 16 is 2\n",
      "row 17 is 1\n",
      "row 17 is 2\n",
      "row 17 is 1\n",
      "row 17 is 2\n",
      "row 18 is 0\n",
      "row 18 is 1\n",
      "row 18 is 1\n",
      "row 18 is 2\n",
      "row 19 is 1\n",
      "row 19 is 2\n",
      "row 19 is 2\n",
      "row 19 is 2\n",
      "row 20 is 2\n",
      "row 20 is 2\n",
      "row 20 is 2\n",
      "row 20 is 2\n",
      "row 21 is 1\n",
      "row 21 is 2\n",
      "row 21 is 2\n",
      "row 21 is 2\n",
      "row 22 is 2\n",
      "row 22 is 1\n",
      "row 22 is 0\n",
      "row 22 is 0\n",
      "row 23 is 0\n",
      "row 23 is 2\n",
      "row 23 is 2\n",
      "row 23 is 2\n",
      "row 24 is 2\n",
      "row 24 is 2\n",
      "row 24 is 1\n",
      "row 24 is 2\n",
      "row 25 is 1\n",
      "row 25 is 2\n",
      "row 25 is 1\n",
      "row 25 is 1\n",
      "row 26 is 1\n",
      "row 26 is 2\n",
      "row 26 is 1\n",
      "row 26 is 2\n",
      "row 27 is 2\n",
      "row 27 is 0\n",
      "row 27 is 1\n",
      "row 27 is 2\n",
      "row 28 is 2\n",
      "row 28 is 1\n",
      "row 28 is 2\n",
      "row 28 is 2\n",
      "row 29 is 2\n",
      "row 29 is 1\n",
      "row 29 is 0\n",
      "row 29 is 2\n",
      "row 30 is 2\n",
      "row 30 is 1\n",
      "row 30 is 2\n",
      "row 30 is 2\n",
      "row 31 is 2\n",
      "row 31 is 2\n",
      "row 31 is 2\n",
      "row 31 is 2\n",
      "row 32 is 1\n",
      "row 32 is 2\n",
      "row 32 is 1\n",
      "row 32 is 2\n",
      "row 33 is 2\n",
      "row 33 is 1\n",
      "row 33 is 2\n",
      "row 33 is 2\n",
      "row 34 is 0\n",
      "row 34 is 2\n",
      "row 34 is 2\n",
      "row 34 is 2\n",
      "row 35 is 2\n",
      "row 35 is 2\n",
      "row 35 is 2\n",
      "row 35 is 0\n",
      "row 36 is 1\n",
      "row 36 is 2\n",
      "row 36 is 2\n",
      "row 36 is 2\n",
      "row 37 is 2\n",
      "row 37 is 2\n",
      "row 37 is 2\n",
      "row 37 is 2\n",
      "row 38 is 0\n",
      "row 38 is 2\n",
      "row 38 is 1\n",
      "row 38 is 2\n",
      "row 39 is 1\n",
      "row 39 is 2\n",
      "row 39 is 2\n",
      "row 39 is 2\n",
      "row 40 is 2\n",
      "row 40 is 1\n",
      "row 40 is 2\n",
      "row 40 is 2\n",
      "row 41 is 1\n",
      "row 41 is 1\n",
      "row 41 is 1\n",
      "row 41 is 2\n",
      "row 42 is 2\n",
      "row 42 is 2\n",
      "row 42 is 2\n",
      "row 42 is 2\n",
      "row 43 is 1\n",
      "row 43 is 2\n",
      "row 43 is 2\n",
      "row 43 is 2\n",
      "row 44 is 1\n",
      "row 44 is 2\n",
      "row 44 is 2\n",
      "row 44 is 2\n",
      "row 45 is 0\n",
      "row 45 is 2\n",
      "row 45 is 1\n",
      "row 45 is 2\n",
      "row 46 is 2\n",
      "row 46 is 2\n",
      "row 46 is 2\n",
      "row 46 is 2\n",
      "row 47 is 0\n",
      "row 47 is 2\n",
      "row 47 is 2\n",
      "row 47 is 2\n",
      "row 48 is 2\n",
      "row 48 is 1\n",
      "row 48 is 2\n",
      "row 48 is 2\n",
      "row 49 is 2\n",
      "row 49 is 1\n",
      "row 49 is 1\n",
      "row 49 is 2\n"
     ]
    }
   ],
   "source": [
    "for i in range(50):\n",
    "    for row in ex1[0:4]:\n",
    "        print(\"row\", i, \"is\",row[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "small = pd.DataFrame(clarks(ex1[0:5]))\n",
    "small[0][1]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def guessUnphased(df, haplotypes):\n",
    "    nSNPs = len(df)\n",
    "    nIndiv = len(df[0])\n",
    "    for i in range(nIndiv):\n",
    "        for j in range(nSNPs):\n",
    "            \n",
    "            if haplotypes[j][2*i] == -1 and haplotypes[j][2*i+1] == -1:\n",
    "                if df[j][i] == 0:\n",
    "                    haplotypes[j][2*i], haplotypes[j][2*i+1] = 0,0\n",
    "                elif df[j][i] == 1:\n",
    "                    haplotypes[j][2*i], haplotypes[j][2*i+1] = 1,0\n",
    "                elif df[j][i] == 2:\n",
    "                    haplotypes[j][2*i], haplotypes[j][2*i+1] = 1,1\n",
    "                    \n",
    "            elif haplotypes[j][2*i] == -1:\n",
    "                if haplotypes[j][2*i+1] == 0:\n",
    "                    if df[j][i] == 0:\n",
    "                        haplotypes[j][2*i] = 0\n",
    "                    elif df[j][i] == 1:\n",
    "                        haplotypes[j][2*i] = 1\n",
    "                    elif df[j][i] == 2:\n",
    "                        # error\n",
    "                        haplotypes[j][i] = 1\n",
    "                elif haplotypes[j][2*i+1] == 1:\n",
    "                    if df[j][i] == 0:\n",
    "                        # error\n",
    "                        haplotypes[j][2*i] = 1\n",
    "                    elif df[j][i] == 1:\n",
    "                        haplotypes[j][2*i] = 0\n",
    "                    elif df[j][i] == 2:\n",
    "                        haplotypes[j][i] = 1\n",
    "            \n",
    "            elif haplotypes[j][2*i+1] == -1:\n",
    "                if haplotypes[j][2*i] == 0:\n",
    "                    if df[j][i] == 0:\n",
    "                        haplotypes[j][2*i+1] = 0\n",
    "                    elif df[j][i] == 1:\n",
    "                        haplotypes[j][2*i+1] = 1\n",
    "                    elif df[j][i] == 2:\n",
    "                        # error\n",
    "                        haplotypes[j][2*i+1] = 1\n",
    "                elif haplotypes[j][2*i] == 1:\n",
    "                    if df[j][i] == 0:\n",
    "                        # error\n",
    "                        haplotypes[j][2*i+1] = 1\n",
    "                    elif df[j][i] == 1:\n",
    "                        haplotypes[j][2*i+1] = 0\n",
    "                    elif df[j][i] == 2:\n",
    "                        haplotypes[j][i+1] = 1\n",
    "                \n",
    "    return haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chunking(genotypes, split_size):\n",
    "    if len(genotypes) % split_size == 0:\n",
    "            numChunks = (int)(len(genotypes)/split_size)\n",
    "    else:\n",
    "        numChunks = (int)(len(genotypes)/split_size) + 1\n",
    "    return numChunks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "#main\n",
    "genots = ex1\n",
    "split_size = 5\n",
    "numChunks = chunking(genots, split_size)\n",
    "haplos = []\n",
    "for i in range(numChunks):\n",
    "    data = genots[i*split_size:(i+1)*split_size]\n",
    "    if i == 0:\n",
    "        haplos = clarks(data)\n",
    "    elif i == numChunks - 1 and len(genots[i*split_size]) > 0:\n",
    "        final = genots[i*split_size:]\n",
    "        final_numSNPs = len(final)\n",
    "        final_numIndivs = len(final[0])\n",
    "        final_haplo_chunk = np.zeros((final_numSNPs, 2*final_numIndivs), dtype = np.int)\n",
    "        final_haplo_chunk.fill(-1)\n",
    "        final_haplo_chunk = guessUnphased(final, final_haplo_chunk)\n",
    "        haplos = np.concatenate((haplos, final_haplo_chunk), axis = 0)\n",
    "    else:\n",
    "        haplos = np.concatenate((haplos, clarks(data)), axis = 0)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "check = pd.DataFrame(haplos)\n",
    "np.savetxt('../test_data_1_my_sol.txt', haplos, fmt='%i', delimiter = ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "sol1 = add_haplos(check[0], check[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "soltrue = pd.read_csv(\"assignment/example_data_1_sol.txt\", header = None, sep = \" \")\n",
    "soltrue = soltrue.astype('int64')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "soltrue0 = add_haplos(soltrue[0],soltrue[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 1,\n",
       " 2,\n",
       " 2,\n",
       " 1,\n",
       " 1,\n",
       " 2,\n",
       " 2,\n",
       " 1,\n",
       " 2,\n",
       " 2,\n",
       " 1,\n",
       " 1,\n",
       " 2,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 1,\n",
       " 0,\n",
       " 1,\n",
       " 2,\n",
       " 1,\n",
       " 2,\n",
       " 0,\n",
       " 2,\n",
       " 1,\n",
       " 1,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 1,\n",
       " 2,\n",
       " 0,\n",
       " 2,\n",
       " 1,\n",
       " 2,\n",
       " 0,\n",
       " 1,\n",
       " 2,\n",
       " 1,\n",
       " 2,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 2,\n",
       " 0,\n",
       " 2,\n",
       " 2]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "soltrue0 == sol1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'iloc'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-149-8b874e4cf891>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msoltrue0\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mex1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'iloc'"
     ]
    }
   ],
   "source": [
    "soltrue0 == ex1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
